src/main/java/pl/javastart/couponcalc/PriceCalculator.java

```diff
@@ -1,13 +1,80 @@
 package pl.javastart.couponcalc;
 
+import java.math.BigDecimal;
+import java.math.RoundingMode;
 import java.util.List;
+import java.util.TreeSet;
 
 public class PriceCalculator {
 
     public double calculatePrice(List<Product> products, List<Coupon> coupons) {
-        // TODO
+        double totalPrice = 0;
+        if (coupons == null && products != null) {
+            totalPrice = calculateTotalPriceWithoutCoupons(products);
+        } else if (coupons != null && products != null) {
+            if (coupons.size() == 1 && coupons.get(0).getCategory() != null) {
+                totalPrice = calculateTotalPriceWithOneCouponWithCategory(coupons, products);
+            } else if (coupons.size() == 1 && coupons.get(0).getCategory() == null) {
+                totalPrice = calculateTotalPriceWithOneCouponWithoutCategory(coupons, products);
+            } else if (coupons.size() > 1) {
+                totalPrice = calculateTotalPriceWithTwoCouponsWithCategory(coupons, products);
+            }
```

Comment:
Na tym poziomie niepotrzebnie rozbijasz to na aż tyle przypadków. Tutaj nie ma znaczenia to, czy jest 1 kupon, czy 10 kuponów. Algorytm zawsze jest taki sam:
- weź kolejny kupon
- oblicz dla niego sumę cen
- sprawdź, czy to jest najniższa suma, a jeżeli tak, to ją zapamiętaj
Jeżeli będzie 1 kupon, to pętla wykona się raz, a jeżeli kuponów będzie 10, to 10 razy.
W tym zadaniu bardziej przydałaby Ci się metoda, która przyjmie listę produktów i kupon i obliczy dla niego sumę, czyli `double getSumWithCoupon(List<Product> products, Coupon coupon)'. Wtedy tutaj możesz mieć pętlę po kuponach, we wspomnianej metodzie drugą pętlę po produktach i to tyle